include/ginkgo/core/base/machine_topology.hpp

```diff
@@ -0,0 +1,399 @@
+    MachineTopology &operator=(MachineTopology &&) = delete;
+    ~MachineTopology() = default;
+
+    /**
+     * Bind the object associated with the ids to a core.
+     *
+     * @param ids  The ids of the object to be bound.
+     */
+    void bind_to_cores(const std::vector<int> &ids) const
+    {
+        hwloc_binding_helper(this->cores_, ids);
+    }
+
+    /**
+     * Bind the object associated with the ids to a Processing unit(PU).
+     *
+     * @param ids  The ids of the object to be bound.
+     */
+    void bind_to_pus(const std::vector<int> &ids) const
+    {
+        hwloc_binding_helper(this->pus_, ids);
+    }
+
+    /**
+     * Get the object of type PU associated with the id.
+     *
+     * @param id  The id of the PU
+     * @return  the PU object struct.
+     */
+    const normal_obj_info *get_pu(size_type id) const
+    {
+        GKO_ENSURE_IN_BOUNDS(id, this->pus_.size());
+        return &this->pus_[id];
+    }
+
+    /**
+     * Get the object of type core associated with the id.
+     *
+     * @param id  The id of the core
+     * @return  the core object struct.
+     */
+    const normal_obj_info *get_core(size_type id) const
+    {
+        GKO_ENSURE_IN_BOUNDS(id, this->cores_.size());
+        return &this->cores_[id];
+    }
+
+    /**
+     * Get the object of type pci device associated with the id.
+     *
+     * @param id  The id of the pci device
+     * @return  the PCI object struct.
+     */
+    const io_obj_info *get_pci_device(size_type id) const
+    {
+        GKO_ENSURE_IN_BOUNDS(id, this->pci_devices_.size());
+        return &this->pci_devices_[id];
+    }
+
+    /**
+     * Get the object of type pci device associated with the PCI bus id.
+     *
+     * @param pci_bus_id  The PCI bus id of the pci device
+     * @return  the PCI object struct.
+     */
+    const io_obj_info *get_pci_device(const std::string &pci_bus_id) const;
+
+    /**
+     * Get the number of PU objects stored in this Topology tree.
+     *
+     * @return  the number of PUs.
+     */
+    size_type get_num_pus() const { return this->pus_.size(); }
+
+    /**
+     * Get the number of core objects stored in this Topology tree.
+     *
+     * @return  the number of cores.
+     */
+    size_type get_num_cores() const { return this->cores_.size(); }
+
+    /**
+     * Get the number of PCI device objects stored in this Topology tree.
+     *
+     * @return  the number of PCI devices.
+     */
+    size_type get_num_pci_devices() const { return this->pci_devices_.size(); }
+
+    /**
+     * Get the number of NUMA objects stored in this Topology tree.
+     *
+     * @return  the number of NUMA objects.
+     */
+    size_type get_num_numas() const { return this->num_numas_; }
+
+protected:
+    /**
+     * Creates a new MachineTopology object.
+     */
+    static std::shared_ptr<MachineTopology> create()
+    {
+        return std::shared_ptr<MachineTopology>(new MachineTopology());
+    }
+
```

Comment:
I don't think you use this function anymore (since you use a local `static` variable instead).
```suggestion
```